astropix.py

```diff
@@ -0,0 +1,701 @@
+        logger.info("Wrote configbits successfully")
+
+
+    # Methods to update the internal variables. Please don't do it manually
+    # This updates the dac config
+    def update_asic_config(self, bias_cfg:dict = None, dac_cfg:dict = None, maskstr:str = None, analog_col:int=None):
+        """
+        Updates and writes confgbits to asic
+
+        bias_cfg:dict - Updates the bias settings. Only needs key/value pairs which need updated
+        dac_cfg:dict - Updates DAC settings. Only needs key/value pairs which need updated
+        """
+        if self.asic_start:
+            if bias_cfg is not None:
+                self.bias_setup.update(bias_cfg)
+            if dac_cfg is not None:
+                self.dac_setup.update(dac_cfg)
+            if maskstr is not None:
+                self._make_digital_mask(maskstr, analog_col)
+            else: 
+                logger.info("update_asic_config() got no argumennts, nothing to do.")
+                return None
+            self.asic_update()
+        else: raise RuntimeError("Asic has not been initalized")
+
+
+    def enable_spi(self):
+        """
+        Starts spi bus. 
+
+        Takes no arguments, returns nothing
+        """
+
+        self.nexys.spi_enable()
+        self.nexys.spi_reset()
+        # Set SPI clockdivider
+        # freq = 100 MHz/spi_clkdiv
+        self.nexys.spi_clkdiv = 255
+        self.nexys.send_routing_cmd()
+        logger.info("SPI ENABLED")
+
+    def close_connection(self):
+        """
+        Terminates the spi bus.
+        Takes no arguments. No returns.
+        """
+        self.nexys.close()
+
+
+################## Voltageboard Methods ############################
+
+# Here we intitalize the 8 DAC voltageboard in slot 4. dacvals are carried over from past 
+# scripts. Default from beam_test.py:
+# Use this: (8, [0, 0, 1.1, 1, 0, 0, 1, 1.035])
+    def init_voltages(self, slot: int = 4, vcal:float = .989, vsupply: float = 2.7, vthreshold:float = None, dacvals: tuple[int, list[float]] = None):
+        """
+        Configures the voltage board
+        No required parameters. No return.
+
+        slot:int = 4 - Position of voltage board
+        vcal:float = 0.908 - Calibration of the voltage rails
+        vsupply = 2.7 - Supply Voltage
+        vthreshold:float = None - ToT threshold value. Takes precedence over dacvals if set. UNITS: mV
+        dacvals:tuple[int, list[float] - vboard dac settings. Must be fully specified if set. 
+        """
+        # The default values to pass to the voltage dac. Last value in list is threshold voltage, default 100mV or 1.1
+        # From nicholas's beam_test.py:
+        # 3 = Vcasc2, 4=BL, 7=Vminuspix, 8=Thpix 
+        default_vdac = (8, [0, 0, 1.1, 1, 0, 0, 1, 1.100])
+        
+        # used to ensure this has been called in the right order:
+        self._voltages_exist = True
+
+        # Set dacvals
+        if dacvals is None:
+            dacvals = default_vdac
```

Comment:
You could probably add a log message here, saying that default settings will be used?